cpp/src/copying/utilities/copying_utils.cuh

```diff
@@ -0,0 +1,163 @@
+  gdf_size_type row_size;         /**< The total size of data that will be copied. */
+
+  //gdf_index_type output_offset; /**< The start index of the output data. */
+};
+
+/**
+ * @brief A helper struct to store information related to the copy of the bitmask.
+ */
+struct bitmask_partition_params {
+  block_type* block_output;             /**< An alias of the output bitmask pointer in 'block_type'. */
+  block_type const* block_input;        /**< An alias of the input bitmask pointer in 'block_type'. */
+
+  gdf_index_type input_offset;          /**< The start index of the input bitmask (pointer in 'block_type'). */
+  block_type rotate_input;              /**< The number of bits that the input bitmask will be rotated or discarded. */
+  double_block_type mask_last;          /**< The mask that will be applied to the last calculated bitmask. */
+
+  gdf_size_type input_block_length;     /**< The size of the input data (gdf_column::size) in 'block_type' units. */
+  gdf_size_type partition_block_length; /**< The size of the interval to be copied in 'block_type' units. */
+
+  //gdf_index_type output_offset;       /**< The start index of the output bitmask (pointer in 'block_type')*/
+  //block_type rotate_output;           /**< The number of bits that the output bitmask will be rotated or discarded. */
+  //double_block_type mask_first;       /**< The mask that will be applied to the first calculated bitmask. */
+};
+
+/**
+ * @brief A CUDA function used to perform the count of the null values.
+ * It counts the bits that are set in the bitmask.
+ */
+__device__ __forceinline__
+void perform_bitmask_null_count(bitmask_partition_params const* params,
+                                gdf_size_type*                  counter,
+                                gdf_index_type const            index) {
+  std::uint32_t bitmask = (std::uint32_t)params->block_output[index];
+  std::uint32_t null_count_value = __popc(bitmask); // Count the number of bits that are set to 1 in a 32 bit integer.
+  atomicAdd(counter, null_count_value);
+}
+
+/**
+ * @brief A CUDA function used to perform the copy of the data.
+ */
+template <typename ColumnType>
+__device__ __forceinline__
+void copy_data(data_partition_params* data_params,
+               ColumnType*            output_data,
+               ColumnType const*      input_data) {
+  // Calculate kernel parameters
+  gdf_size_type row_index = threadIdx.x + blockIdx.x * blockDim.x;
+  gdf_size_type row_step = blockDim.x * gridDim.x;
+
+  // Perform the copying operation
+  while (row_index < data_params->row_size) {
+    output_data[row_index] = input_data[data_params->input_offset + row_index];
+    row_index += row_step;
+  }
+}
+
+/**
+ * @brief A CUDA function used to perform the copy of the bitmask.
+ * The function reads a block of size 'double_block_type' from the input
+ * bitmask and process (apply rotation and logical operators) in order to
```

Comment:
This is to copy a slice of a bitmask. If the slice is from element 10 to element 40, element 10 corresponds to bit 3 of the second byte, that bit needs to become bit 0. So we are reading two adjacent blocks and bitshifting them together. To then write one block

Reply:
I see. Please add this example to the docs, it's helpful.

Reply:
Thinking about this more... If the slice spans more than two bit_mask_t elements, then presumably you will be using as many threads as there are bit_mask_t elements int he output to do the copy. Every thread is reading two elements and writing one. Thus the kernel is using 50% more bandwidth than it needs to. It would be more efficient to stage via shared memory.  Each thread loads *one* word from global and stores it in shared memory. Then sync threads. Then threads load two words from shared memory to compute a single output word (just as in this function), and they then write that output to global memory. 

Since shared memory bandwidth is much higher than global memory bandwidth, this should be quite a bit faster.

Reply:
I would like to revisit this later for performance